### PR TITLE
api: add TypeTmpfs to api/types/mount

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**\- What I did**
added `TypeTmpfs` to `api/types/mount`.

TODO: migrate Swarmkit and CLI to use the new mount API

**\- How I did it**

**\- How to verify it**
`TESTFLAGS='-check.f TestContainersApiCreateMounts' make test-integration-cli`

**\- Description for the changelog**

api: add TypeTmpfs to api/types/mount

**\- A picture of a cute animal (not mandatory but encouraged)**
penguins

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
CC @cpuguy83

File: api/types/mount/mount.go [hunk 1]
```diff
@@ -56,3 +61,11 @@ type Driver struct {
 	Name    string            `json:",omitempty"`
 	Options map[string]string `json:",omitempty"`
 }
+
+// TmpfsOptions defines options specific to mounts of type "tmpfs".
+type TmpfsOptions struct {
+	// RawOptions is the raw string passed to mount(2).
+	// e.g. "rw,noexec,nosuid,size=65536k"
+	// RawOptions cannot contain "ro,", currently.
+	RawOptions string `json:",omitempty"`
```
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: It is easy to support readonly tmpfs, but needs consideration for `m.ReadOnly==true && m.TmpfsOptions.RawOptions.contains("rw")`.
Not sure actually there have been people using readonly tmpfs (theoretically it can be used for hiding some files without deletion)

**Comment by justincormack** [comment]
cc @stevvooe

File: container/container_unix.go [hunk 2]
```diff
@@ -415,6 +416,19 @@ func (container *Container) TmpfsMounts() []Mount {
 			Data:        data,
 		})
 	}
+	for _, mnt := range container.HostConfig.Mounts {
+		if mnt.Type == mounttypes.TypeTmpfs {
```
Comment (cpuguy83) [inline container/container_unix.go]: We should have these added to container.Mountpoints now that we have a proper "Type" field.... it would also be helpful to have `HostConfig.Tmpfs` converted into this format like we are doing (daemon side) for stuff in `Binds` and `Volumes`.

File: api/types/mount/mount.go [hunk 3]
```diff
@@ -8,17 +8,21 @@ const (
 	TypeBind Type = "bind"
 	// TypeVolume VOLUME
 	TypeVolume Type = "volume"
+	// TypeTmpfs TMPFS
```
Comment (cpuguy83) [inline api/types/mount/mount.go]: :) We should just group these into a single comment above the `const` block describing what they are for rather than having these redundant doc strings.

File: api/types/mount/mount.go [hunk 4]
```diff
@@ -56,3 +60,12 @@ type Driver struct {
...
 	Options map[string]string `json:",omitempty"`
 }
+
+// TmpfsOptions defines options specific to mounts of type "tmpfs".
+type TmpfsOptions struct {
+	// RawOptions is the raw string passed to mount(2).
+	// e.g. "rw,noexec,nosuid,size=65536k"
+	// RawOptions can contain "ro" or "rw" but needs to be consistent with
+	// Mount.ReadOnly
+	RawOptions string `json:",omitempty"`
```
Comment (cpuguy83) [inline api/types/mount/mount.go]: I think swarmkit is using `Mode` here.

```go
type Mount_TmpfsOptions struct {
	// Size sets the size of the tmpfs, in bytes.
	//
	// This will be converted to an operating system specific value
	// depending on the host. For example, on linux, it will be convered to
	// use a 'k', 'm' or 'g' syntax. BSD, though not widely supported with
	// docker, uses a straight byte value.
	//
	// Percentages are not supported.
	SizeBytes int64 `protobuf:"varint,1,opt,name=size_bytes,json=sizeBytes,proto3" json:"size_bytes,omitempty"`
	// Mode of the tmpfs upon creation
	Mode os.FileMode `protobuf:"varint,2,opt,name=mode,proto3,customtype=os.FileMode" json:"mode"`
}
```

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: integration-cli/docker_api_containers_test.go [hunk 5]
```diff
@@ -1547,6 +1547,12 @@ func (s *DockerSuite) TestContainersApiCreateMountsValidation(c *check.C) {
 		{cfg{Image: "busybox", HostConfig: hc{Mounts: []m{{Type: "volume"}}}}, http.StatusBadRequest, "Target must not be empty"},
 		{cfg{Image: "busybox", HostConfig: hc{Mounts: []m{{Type: "volume", Source: "hello", Target: destPath}}}}, http.StatusCreated, ""},
 		{cfg{Image: "busybox", HostConfig: hc{Mounts: []m{{Type: "volume", Source: "hello2", Target: destPath, VolumeOptions: &mounttypes.VolumeOptions{DriverConfig: &mounttypes.Driver{Name: "local"}}}}}}, http.StatusCreated, ""},
+		{cfg{Image: "busybox", HostConfig: hc{Mounts: []m{{Type: "tmpfs", Target: destPath}}}}, http.StatusCreated, ""},
```
Comment (stevvooe) [inline integration-cli/docker_api_containers_test.go]: Could we format these to make them readable?

File: integration-cli/docker_api_containers_test.go [hunk 6]
```diff
@@ -1730,3 +1736,48 @@ func (s *DockerSuite) TestContainersApiCreateMountsCreate(c *check.C) {
 		}
 	}
 }
+
+func (s *DockerSuite) TestContainersApiCreateMountsTmpfs(c *check.C) {
+	testRequires(c, DaemonIsLinux)
```
Comment (stevvooe) [inline integration-cli/docker_api_containers_test.go]: Is tmpfs linux only? Windows does have ramdisk.

File: integration-cli/docker_api_containers_test.go [hunk 7]
```diff
@@ -1730,3 +1736,48 @@ func (s *DockerSuite) TestContainersApiCreateMountsCreate(c *check.C) {
 		}
 	}
 }
+
+func (s *DockerSuite) TestContainersApiCreateMountsTmpfs(c *check.C) {
```
Comment (stevvooe) [inline integration-cli/docker_api_containers_test.go]: `TestContainersAPICreateMountsTmpfs`.

Please follow Go's naming conventions.

File: api/types/mount/mount.go (hunk 4 above)
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: ok, I'm going to change the struct definition to like this:
```go
// TmpfsOptions defines options specific to mounts of type "tmpfs".                                                                      
type TmpfsOptions struct {
        // Size sets the size of the tmpfs, in bytes.                                                                                    
        //                                                                                                                               
        // This will be converted to an operating system specific value                                                                  
        // depending on the host. For example, on linux, it will be convered to                                                          
        // use a 'k', 'm' or 'g' syntax. BSD, though not widely supported with                                                           
        // docker, uses a straight byte value.                                                                                           
        //                                                                                                                               
        // Percentages are not supported.                                                                                                
        SizeBytes int64
        // Mode of the tmpfs upon creation                                                                                               
        Mode os.FileMode

        // RawOptions is the raw string passed to mount(2).                                                                              
        // e.g. "rw,noexec,nosuid,size=65536k"                                                                                           
        //                                                                                                                               
        // RawOptions is exclusive to other TmpfsOptions fields.                                                                         
        // If RawOptions has a non-empty value, other fields must have empty values.                                                     
        //                                                                                                                               
        // RawOptions can contain "ro" or "rw" but needs to be consistent with                                                           
        // Mount.ReadOnly                                                                                                                
        RawOptions string `json:",omitempty"`
}
```

I think we need to keep RawOptions for compatibility, WDYT?

File: integration-cli/docker_api_containers_test.go (hunk 6 above)
Comment (AkihiroSuda) [inline integration-cli/docker_api_containers_test.go]: AFAIK yes, linux only in the current implementation?
https://github.com/docker/docker/blob/b8265e55508db99aea632033b2f5008f921b11e2/container/container_windows.go#L89-L93

IMO implementation for Windows should be another PR, but I understand we need to design the API with consideration for Windows in this PR.

CC @jhowardmsft

File: integration-cli/docker_api_containers_test.go (hunk 7 above)
Comment (AkihiroSuda) [inline integration-cli/docker_api_containers_test.go]: It seems there is some project-local naming convention that encourages `Test.*Api.*`?
If not, perhaps we need to fix that in separate PR?

```console
$ grep "func (.*Suite) Test.*Api" integration-cli/*.go | wc -l
158
$ grep "func (.*Suite) Test.*API" integration-cli/*.go | wc -l
2
```

File: container/container_unix.go (hunk 2 above)
Comment (AkihiroSuda) [inline container/container_unix.go]: done

File: api/types/mount/mount.go (hunk 3 above)
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: done

File: integration-cli/docker_api_containers_test.go (hunk 5 above)
Comment (AkihiroSuda) [inline integration-cli/docker_api_containers_test.go]: done

**Comment by AkihiroSuda** [comment]
updated PR, please review again if CI greeen?
@cpuguy83 @stevvooe

File: daemon/volumes.go [hunk 8]
```diff
@@ -197,13 +196,36 @@ func (daemon *Daemon) registerMountPoints(container *container.Container, hostCo
...
+		}
+		mountPoints[mp.Destination] = mp
+	}
+
+	// 4. Set up tmpfs
+	for dest, data := range hostConfig.Tmpfs {
+		spec := mounttypes.Mount{
+			Type:   mounttypes.TypeTmpfs,
+			Target: dest,
+			// FIXME: set ReadOnly to true depending on the data
```
Comment (AkihiroSuda) [inline daemon/volumes.go]: forgot to fix this, now fixing..

File: integration-cli/docker_api_containers_test.go (hunk 7 above)
Comment (thaJeztah) [inline integration-cli/docker_api_containers_test.go]: Agreed; better done in a separate PR if we want to change that

**Comment by AkihiroSuda** [comment]
Failure in windowsRS1 is trivial, I'll make sure it is skipped on Windows
```console
00:48:59.244 ----------------------------------------------------------------------
00:48:59.245 FAIL: docker_api_containers_test.go:1525: DockerSuite.TestContainersApiCreateMountsValidation
00:48:59.245 
00:48:59.245 case 0
00:48:59.245 case 1
00:48:59.245 case 2
00:48:59.245 case 3
00:48:59.245 case 4
00:48:59.245 case 5
00:48:59.245 case 6
00:48:59.245 case 7
00:48:59.245 docker_api_containers_test.go:1711:
00:48:59.245     c.Assert(status, checker.Equals, x.status, check.Commentf("%s\n%v", string(b), cases[i].config))
00:48:59.245 ... obtained int = 400
00:48:59.245 ... expected int = 201
00:48:59.246 ... {"message":"invalid mount config for type \"tmpfs\": Windows does not support tmpfs"}
00:48:59.246 
00:48:59.246 {busybox {[{tmpfs  c:\foo false <nil> <nil> <nil>}]}}
00:48:59.246 
00:48:59.395 
00:48:59.395 ----------------------------------------------------------------------
```

**Review by stevvooe — CHANGES_REQUESTED** [review]

File: integration-cli/docker_api_containers_test.go (hunk 7 above)
Comment (stevvooe) [inline integration-cli/docker_api_containers_test.go]: @thaJeztah This is a new test. It should follow our coding standards.

Demonstrations of others not following the standards does not make it okay to not follow the standards.

File: container/container_unix.go [hunk 9]
```diff
@@ -415,6 +416,19 @@ func (container *Container) TmpfsMounts() []Mount {
...
+			data := ""
+			if mnt.TmpfsOptions != nil {
+				data = mnt.TmpfsOptions.RawOptions
+			}
+			mounts = append(mounts, Mount{
+				Source:      "tmpfs",
```
Comment (AkihiroSuda) [inline container/container_unix.go]: It is from the original source https://github.com/docker/docker/pull/26837/files#diff-5dfdabc4cd5100a2aee949c12d2afe6cL413

File: integration-cli/docker_api_containers_test.go (hunk 7 above)
Comment (AkihiroSuda) [inline integration-cli/docker_api_containers_test.go]: ok, for existing tests, opened PR https://github.com/docker/docker/pull/26967

**Comment by AkihiroSuda** [comment]
updated PR: removed `RawOptions` from `api/types/mount.TmpfsOptions`

File: api/types/mount/mount.go [hunk 10]
```diff
@@ -1,24 +1,30 @@
...
+
 // Type represents the type of a mount.
 type Type string
 
+// Type constants
 const (
-	// TypeBind BIND
-	TypeBind Type = "bind"
-	// TypeVolume VOLUME
+	TypeBind   Type = "bind"
```
Comment (stevvooe) [inline api/types/mount/mount.go]: Rather than remove documentation, let's actually add it in. You can copy from swarmkit.

File: api/types/mount/mount.go [hunk 11]
```diff
@@ -1,24 +1,30 @@
...
 	TypeVolume Type = "volume"
+	TypeTmpfs  Type = "tmpfs"
 )
 
 // Mount represents a mount (volume).
 type Mount struct {
-	Type     Type   `json:",omitempty"`
+	Type Type `json:",omitempty"`
+	// Source is not supported in TMPFS (must be an empty value)
```
Comment (stevvooe) [inline api/types/mount/mount.go]: There is more to `Source` than just that. This is incomplete.

File: api/types/mount/mount.go [hunk 12]
```diff
@@ -56,3 +62,24 @@ type Driver struct {
...
+	// depending on the host. For example, on linux, it will be convered to
+	// use a 'k', 'm' or 'g' syntax. BSD, though not widely supported with
+	// docker, uses a straight byte value.
+	//
+	// Percentages are not supported.
+	SizeBytes int64
+	// Mode of the tmpfs upon creation
+	Mode os.FileMode
+
+	UID          uint32
```
Comment (stevvooe) [inline api/types/mount/mount.go]: How can any of this possibly be supported if we don't have support in swarmkit? https://github.com/docker/swarmkit/blob/master/api/types.proto#L191

File: volume/volume_unix.go [hunk 13]
```diff
@@ -136,3 +137,86 @@ func clean(p string) string {
...
+			return "", fmt.Errorf("Blocks is not supported on %s", os)
+		}
+		if opt.SizeBytes != 0 {
+			return "", fmt.Errorf("exclusive: Blocks vs SizeBytes")
+		}
+		rawOpts = append(rawOpts, fmt.Sprintf("nr_blocks=%d", opt.Blocks))
+	}
+	if opt.INodes != 0 {
+		switch os {
+		case "linux":
```
Comment (stevvooe) [inline volume/volume_unix.go]: Use build tags to do this.

File: volume/volume_unix_test.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+package volume
+
+import (
+	"testing"
+
+	mounttypes "github.com/docker/docker/api/types/mount"
+)
+
+func TestRawTmpfsOptions(t *testing.T) {
```
Comment (stevvooe) [inline volume/volume_unix_test.go]: There is no more raw...

**Review by stevvooe — CHANGES_REQUESTED** [review]

File: api/types/mount/mount.go (hunk 12 above)
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: they are for non-swarm containers, and imo they are still needed for compatibility; old tmpfs allowed these options

WDYT?

File: volume/volume_unix.go [hunk 15]
```diff
@@ -136,3 +137,86 @@ func clean(p string) string {
...
+				size = size / r.divisor
+				suffix = r.suffix
+				break
+			}
+		}
+
+		rawOpts = append(rawOpts, fmt.Sprintf("size=%d%s", size, suffix))
+	}
+	if opt.UID != 0 {
+		rawOpts = append(rawOpts, fmt.Sprintf("uid=%d", opt.UID))
```
Comment (AkihiroSuda) [inline volume/volume_unix.go]: TODO
Check nil instead of zero

File: api/types/mount/mount.go (hunk 12 above)
Comment (stevvooe) [inline api/types/mount/mount.go]: These were not supported with tmpfs previously. Do not add them in this PR. You can have a similar comment as I have in swarmkit, but please don't add things that aren't actually supportable.

Comment (AkihiroSuda) [inline api/types/mount/mount.go]: It seems supported it in Docker 1.12.2, WDYT?
```console
$ docker run -it --rm --tmpfs /foo:uid=4242,nr_inodes=3 busybox
/ # ls -ld /foo
drwxrwxrwt    2 4242     root            40 Oct 14 05:10 /foo
/ # touch /foo/1
/ # touch /foo/2
/ # touch /foo/3
touch: /foo/3: No space left on device
```

Comment (stevvooe) [inline api/types/mount/mount.go]: They are not currently supported for swarmkit. I am not sure what to say about docker run. These options are in no way portable.

File: api/types/mount/mount.go [hunk 16]
```diff
@@ -56,3 +62,24 @@ type Driver struct {
...
+	GID          uint32
+	Blocks       int64
+	INodes       int64
+	MemoryPolicy string
```
Comment (stevvooe) [inline api/types/mount/mount.go]: What is the contents of `MemoryPolicy`?

File: api/types/mount/mount.go (hunk 4 above)
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: now TmpfsOptions is identical to the corresponding one of Swarmkit

File: api/types/mount/mount.go (hunk 10 above)
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: done

File: api/types/mount/mount.go (hunk 11 above)
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: done

File: volume/volume_unix.go (hunk 13 above)
Comment (AkihiroSuda) [inline volume/volume_unix.go]: removed the code itself

File: api/types/mount/mount.go (hunk 16 above)
Comment (AkihiroSuda) [inline api/types/mount/mount.go]: I removed it in the latest commit, but it meant `mpol` for Linux tmpfs

https://linux.die.net/man/8/mount
> mpol=[default|prefer:Node|bind:NodeList|interleave|interleave:NodeList]

**Comment by AkihiroSuda** [comment]
Updated PR @stevvooe @cpuguy83

File: volume/volume.go [hunk 17]
```diff
@@ -104,8 +110,9 @@ type MountPoint struct {
 	CopyData bool `json:"-"`
 	// ID is the opaque ID used to pass to the volume driver.
 	// This should be set by calls to `Mount` and unset by calls to `Unmount`
-	ID   string `json:",omitempty"`
-	Spec mounttypes.Mount
+	ID    string `json:",omitempty"`
+	Tmpfs *Tmpfs `json:"-"`
```
Comment (stevvooe) [inline volume/volume.go]: If we have the spec, why do we need to include the `*Tmpfs` type here, as well?

Comment (AkihiroSuda) [inline volume/volume.go]: Unlike the spec structure, `*Tmpfs` here contains `RawOptions` string, which is needed to the support for the old tmpfs in `docker run`.

Comment (stevvooe) [inline volume/volume.go]: @AkihiroSuda Ok, but let's not fork these into two paths. If there are options we need for docker run support, then let's add them. For now, let's get swarm-mode support for `Tmpfs` without introducing another half-baked type that has to be supported forever.

Comment (stevvooe) [inline volume/volume.go]: Put differently, let's not cut `docker run` over to using the mounts for tmpfs until there is parity.

I think we need to understand the use case of `uid/gid` and `nr_blocks` to better support these across operating systems. Right now, these are extremely non-portable.

File: integration-cli/docker_api_containers_test.go [hunk 18]
```diff
@@ -1759,3 +1879,41 @@ func (s *DockerSuite) TestContainersAPICreateMountsCreate(c *check.C) {
...
+
+func (s *DockerSuite) TestContainersAPICreateMountsTmpfs(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+	type testCase struct {
+		cfg             map[string]interface{}
+		expectedOptions []string
+	}
+	target := "/foo"
+	cases := []testCase{
+		{map[string]interface{}{
```
Comment (stevvooe) [inline integration-cli/docker_api_containers_test.go]: Please label the fields for struct literals.

File: volume/volume.go [hunk 19]
```diff
@@ -290,6 +297,14 @@ func ParseMountSpec(cfg mounttypes.Mount, options ...func(*validateOpts)) (*Moun
 				mp.Propagation = cfg.BindOptions.Propagation
 			}
 		}
+	case mounttypes.TypeTmpfs:
+		rawOptions, err := convertTmpfsOptions(cfg.TmpfsOptions)
```
Comment (stevvooe) [inline volume/volume.go]: Why drop type safety here?

File: volume/volume.go (hunk 17 above)
Comment (AkihiroSuda) [inline volume/volume.go]: For passing the spec to the `mount(2)` syscall, we need to compile the typed spec struct to string in somewhere.

I can come up with an alternative that defers such compilation, but I think it is complicated

alternative `volume/volume.go`:
```go
import (
        mounttypes "github.com/docker/docker/api/types/mount"
)

type Tmpfs struct {
  Options *mounttypes.TmpfsOptions // typed spec, for `--mount`
  RawOptions *string // raw string, for old `--tmpfs`
}
```

alternative `container/container_unix.go`:
```go
func (container *Container) TmpfsMounts() ([]Mount, error) {
	var mounts []Mount
	for dest, mnt := range container.MountPoints {
		if mnt.Type == mounttypes.TypeTmpfs {
			data := ""
			if mnt.Tmpfs != nil {
				if mnt.Tmpfs.RawOptions == nil && if mnt.Tmpfs.Options == nil {
					return nil, fmt.Errorf("some error")
				}
				if mnt.Tmpfs.RawOptions != nil && if mnt.Tmpfs.Options != nil {
					return nil, fmt.Errorf("some error")
				}			
			        if mnt.Tmpfs.RawOptions != nil {
					data = mnt.Tmpfs.RawOptions
				} else if mnt.Tmpfs.Options != nil {
					data = ConvertTmpfsOptions(mnt.Tmpfs.Options)
				}
			}
			mounts = append(mounts, Mount{
				Source:      "tmpfs",
				Destination: dest,
				Data:        data,
			})
		}
	}

	return mounts, nil
}
```

File: volume/volume.go (hunk 19 above)
Comment (AkihiroSuda) [inline volume/volume.go]: Please refer to https://github.com/docker/docker/pull/26837#discussion_r84215734

File: integration-cli/docker_api_containers_test.go (hunk 18 above)
Comment (AkihiroSuda) [inline integration-cli/docker_api_containers_test.go]: done

File: volume/volume.go (hunk 17 above)
Comment (AkihiroSuda) [inline volume/volume.go]: ok, updated PR.
Now the old `HostConfig.Tmpfs` is not converted to the new typed mount spec.
The typed mount spec is compiled to the raw mount option string when it is really needed.

i.e. `container/container_unix.go`

```go
// TmpfsMounts returns the list of tmpfs mounts
func (container *Container) TmpfsMounts() ([]Mount, error) {
	var mounts []Mount
	for dest, data := range container.HostConfig.Tmpfs {
		mounts = append(mounts, Mount{
			Source:      "tmpfs",
			Destination: dest,
			Data:        data,
		})
	}
	for dest, mnt := range container.MountPoints {
		if mnt.Type == mounttypes.TypeTmpfs {
			data, err := volume.ConvertTmpfsOptions(mnt.Spec.TmpfsOptions)
			if err != nil {
				return nil, err
			}
			mounts = append(mounts, Mount{
				Source:      "tmpfs",
				Destination: dest,
				Data:        data,
			})
		}
	}
	return mounts, nil
}
```

**Comment by AkihiroSuda** [comment]
updated PR https://github.com/docker/docker/pull/26837#discussion_r84838378
Please look into this again (if CI green)? @stevvooe @cpuguy83

File: volume/validate.go [hunk 20]
```diff
@@ -87,6 +87,13 @@ func validateMountConfig(mnt *mount.Mount, options ...func(*validateOpts)) error
 				return &errMountConfig{mnt, err}
 			}
 		}
+	case mount.TypeTmpfs:
+		if len(mnt.Source) != 0 {
+			return &errMountConfig{mnt, fmt.Errorf("must not set Source")}
```
Comment (cpuguy83) [inline volume/validate.go]: Please use `errExtraField` instead of `fmt.Errorf`

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Pretty happy with this, just one minor change and I think we'll be ready to roll.

**Review by stevvooe — APPROVED** [review]

**Comment by stevvooe** [comment]
The fact that `Mount.Data` exists is disappointing but I don't see how we can remove it.

Thanks for the solid PR!

**Comment by thaJeztah** [comment]
ping @AkihiroSuda can you address @cpuguy83's nit (https://github.com/docker/docker/pull/26837#discussion_r84901831), then it looks like we're ready to go

**Comment by AkihiroSuda** [comment]
@thaJeztah done

**Comment by AkihiroSuda** [comment]
updated docs/reference/api as well

**Review by cpuguy83 — APPROVED** [review]
LGTM

Moving to docs review

File: docs/reference/api/docker_remote_api_v1.25.md [hunk 21]
```diff
@@ -525,6 +526,9 @@ Create a container
             - **DriverConfig** – Map of driver-specific options.
               - **Name** - Name of the driver to use to create the volume.
               - **Options** - key/value map of driver specific options.
+        - **TmpfsOptions** – Optional configuration for the `tmpfs` type.
+            - **SizeBytes** – The size for the tmpfs mount in bytes.
+            - **Mode** – The permission mode for the tmpfs mount in an integer.
```
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.25.md]: Not for this PR, but I think we should update the JSON example with some mounts

**Review by thaJeztah — APPROVED** [review]
docs LGTM, thanks so much @AkihiroSuda!

**Review by vdemeester — APPROVED** [review]
LGTM 🐸
